### PR TITLE
sys/net/nanocoap: refactor to remove work-buffer requirement

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

Nanocoap would always require a work buffer of CoAP header + payload block size for (blockwise) requests, even though this is then handed to the user via a callback and has to be copied again.

Avoid the need for this additional buffer and copying by using the `sock_udp_recv_buf()` API and hand the user callback a pointer to network stack memory directly without the need to copy it to a separate temporary buffer.

### Testing procedure

<details><summary>examples/suit_update</summary>

```
suit: received URL: "coap://[2001:db8::1]/fw/same54-xpro/suit_update-riot.suit_signed.latest.bin "
suit_coap: trigger received
suit_coap: downloading "coap://[2001:db8::1]/fw/same54-xpro/suit_update-riot.suit_signed.latest.bin"
suit_coap: got manifest with size 495
suit: verifying manifest signature
suit: validated manifest version
)riotboot_hdr_validate: riotboot_hdr magic number invalid
Manifest seq_no: 1648056359, highest available: 1648056326
suit: validated sequence number
)Formatted component name: 
Comparing manifest offset 4000 with other slot offset
Comparing manifest offset 82000 with other slot offset
validating vendor ID
Comparing 547d0d74-6d3a-5a92-9662-4881afd9407b to 547d0d74-6d3a-5a92-9662-4881afd9407b from manifest
validating vendor ID: OK
validating class id
Comparing 50244518-6a7c-5ce7-932b-88b318336c82 to 50244518-6a7c-5ce7-932b-88b318336c82 from manifest
validating class id: OK
Comparing manifest offset 4000 with other slot offset
Comparing manifest offset 82000 with other slot offset
SUIT policy check OK.
Formatted component name: 
riotboot_flashwrite: initializing update to target slot 1
Fetching firmware |█████████████████████████| 100%
Finalizing payload store
Verifying image digest
Starting digest verification against image
Install correct payload
Verifying image digest
Starting digest verification against image
Install correct payload
Image magic_number: 0x544f4952
Image Version: 0x623b5827
Image start address: 0x00082400
Header chksum: 0x2cd27c0c

suit_coap: rebooting...
----> ethos: hello received
NETOPT_RX_END_IRQ not implemented by driver
gnrc_uhcpc: Using 4 as border interface and 0 as wireless interface.
gnrc_uhcpc: only one interface found, skipping setup.
main(): This is RIOT! (Version: 2022.04-devel-929-g0fd0e9-nanocoap-rework)
RIOT SUIT update example application
Running from slot 1
Image magic_number: 0x544f4952
Image Version: 0x623b5827
Image start address: 0x00082400
Header chksum: 0x2cd27c0c

suit_coap: started.
Starting the shell
> 
```
</details>

### Issues/PRs references

Undoes the API change from #17474 that requires a work buffer to be supplied.
Before #17474 the work buffer was internal and on the stack. This was a problem for large block sizes, so that PR moved the issue of allocating a work buffer to the user.

Now this PR does away with the need for the work buffer to include the payload, it is only needed for the CoAP header.
This lets us restore the original API again.

### CODE REVIEW

File: sys/include/net/nanocoap.h [hunk 1]
```diff
@@ -157,17 +157,9 @@ extern "C" {
  * @brief    Maximum length of a CoAP header for a blockwise message
  */
 #ifndef CONFIG_NANOCOAP_BLOCK_HEADER_MAX
-#define CONFIG_NANOCOAP_BLOCK_HEADER_MAX   (64)
+#define CONFIG_NANOCOAP_BLOCK_HEADER_MAX   (80)
```
Comment (maribu) [inline sys/include/net/nanocoap.h:160]: Hmm, is there some reasoning behind this number? If so, it would be helpful to add a short comment explaining that reasoning.

File: sys/net/application_layer/nanocoap/sock.c [hunk 2]
```diff
@@ -102,41 +136,100 @@ ssize_t nanocoap_sock_request(nanocoap_sock_t *sock, coap_pkt_t *pkt, size_t len
                 continue;
             }
 
-            break;
+            switch (coap_get_type(pkt)) {
+            case COAP_TYPE_RST:
+                /* TODO: handle different? */
+                return -EBADMSG;
+            case COAP_TYPE_CON:
+                /* TODO: send ACK response */
```
Comment (maribu) [inline sys/net/application_layer/nanocoap/sock.c:144]: Without the ACK the communication partner will re-transmit the CON for roughly 5 minutes or so with default parameters. That really should be done to allow the partner to free the resources earlier and put less load on the network.

File: sys/net/application_layer/nanocoap/sock.c [hunk 3]
```diff
@@ -102,41 +136,100 @@ ssize_t nanocoap_sock_request(nanocoap_sock_t *sock, coap_pkt_t *pkt, size_t len
...
+                /* fall-through */
+            case COAP_TYPE_NON:
+                if (pkt->payload_len == 0) {
+                    return 0;
+                }
+                /* fall-through */
+            case COAP_TYPE_ACK:
+                if (pkt->payload_len == 0 && response) {
+                    state = STATE_AWAIT_RESPONSE;
+                    /* TODO: adjust timeout */
```
Comment (maribu) [inline sys/net/application_layer/nanocoap/sock.c:154]: That's difficult to do. A server really only should reply with an empty ACK if waiting for the content of the reply would not be sensible (e.g. because the client is bound to assume a timeout and perform needless retransmissions). The standard only has this to say about what the timeout should be for the separate response:

> Note also that, while the CoAP protocol itself
      does not make any specific demands here, there is an expectation
      that the response will come within a time frame that is reasonable
      from an application point of view.

File: sys/include/net/nanocoap_sock.h [hunk 4]
```diff
@@ -277,6 +273,24 @@ ssize_t nanocoap_get_blockwise_url_to_buf(const char *url,
...
+ * @param[in,out]   pkt     Packet struct containing the request. Is reused for
+ *                          the response
+ * @param[in]       cb      Callback executed for response packet
+ * @param[in]       arg     Optional callback argumnent
+ *
+ * @returns     length of response on success
+ * @returns     <0 on error
+ */
+ssize_t nanocoap_sock_request_cb(sock_udp_t *sock, coap_pkt_t *pkt,
+                                 coap_request_cb_t cb, void *arg);
```
Comment (maribu) [inline sys/include/net/nanocoap_sock.h:292]: In the implementation I can see that the response callback is not executed when the response has no payload. I'm not sure if that is a good idea, a caller likely still wants to evaluate the status code of the response.

E.g. consider a `2.03 Valid` reply without payload that contains an `Etag` Option. In this case the response would have a payload from a high level point of view: The client included one or more Etag Options in the request to tell the server which states of the resource the client has a cached payload for, so that the server can skip transmitting the payload and just tell which of the provided Etags is now valid.

File: sys/net/application_layer/nanocoap/sock.c [hunk 5]
```diff
@@ -102,41 +136,100 @@ ssize_t nanocoap_sock_request(nanocoap_sock_t *sock, coap_pkt_t *pkt, size_t len
...
+    memcpy(buf->iov_base, pkt->hdr, pkt_len);
+
+    pkt->hdr = buf->iov_base;
+    pkt->token  = (uint8_t*)pkt->hdr + sizeof(coap_hdr_t);
+    pkt->payload = (uint8_t*)pkt->hdr + (pkt_len - pkt->payload_len);
+
+    return pkt_len;
+}
+
+ssize_t nanocoap_sock_request(sock_udp_t *sock, coap_pkt_t *pkt, size_t len)
```
Comment (maribu) [inline sys/net/application_layer/nanocoap/sock.c:185]: I'm aware that this function is preexisting, but isn't the length already implicitly known due to the data provided by `pkt`? This looks to me like inconsistent interpretations of the pkt length could be introduced via the `len` parameter, if the caller doesn't keep `pkt->payload_len` in sync.

I have the fear that this can result in fun bugs :/

**Review by maribu — COMMENTED** [review]
Some comments inline

File: sys/net/application_layer/nanocoap/sock.c (hunk 3 above)
Comment (benpicco) [reply sys/net/application_layer/nanocoap/sock.c:154]: So it's best to leave this as is?

File: sys/net/application_layer/nanocoap/sock.c (hunk 2 above)
Comment (benpicco) [reply sys/net/application_layer/nanocoap/sock.c:144]: Is sending an empty ACK with code 2.03 the right thing to do here?

File: sys/net/application_layer/nanocoap/sock.c (hunk 5 above)
Comment (benpicco) [reply sys/net/application_layer/nanocoap/sock.c:185]: > isn't the length already implicitly known due to the data provided by `pkt`?

No `pkt` only contains the length of the request, but the idea is to re-use the request buffer for the response - whose length needs to be supplied separately.

(Yea I know… a separate buffer, argument that could just so happen to be an alias for the pkg buffer would be the cleaner solution here.)

File: sys/net/application_layer/nanocoap/sock.c (hunk 2 above)
Comment (maribu) [reply sys/net/application_layer/nanocoap/sock.c]: An empty ACK contains no code, as it is is only used when a meaningful reply is not yet possible.

I'm currently on my phone. If I recall correctly it is about a separate response. To that, one would react with an empty ACK (without code), as what was received is already the response (and there is no sensible reply to a response in the request-response model of CoAP). So the empty ACK in that case only is used to cease retransmissions.

File: sys/net/application_layer/nanocoap/sock.c (hunk 3 above)
Comment (maribu) [reply sys/net/application_layer/nanocoap/sock.c]: Likely one should, as the standard indicates, shift the responsibility to the application code. But it would be a pain in the ass to let all callers provide a timeout as parameter even if they do not expect a separate respone at all. But semantic-wise, I guess that would be the way to go.

One could also provide a compile time macro with sone sensible default and let the application adjust it as needed. But anyone who has a use case where different timeouts for different requests are needed will start throwing rotten tomatos at us.

File: sys/net/application_layer/nanocoap/sock.c [hunk 6]
```diff
@@ -102,41 +147,91 @@ ssize_t nanocoap_sock_request(nanocoap_sock_t *sock, coap_pkt_t *pkt, size_t len
...
+            case COAP_TYPE_RST:
+                /* TODO: handle different? */
+                return -EBADMSG;
+            case COAP_TYPE_CON:
+                _send_ack(sock, pkt);
+                /* fall-through */
+            case COAP_TYPE_NON:
+            case COAP_TYPE_ACK:
+                /* call user callback */
+                res = cb(arg, pkt);
```
Comment (maribu) [inline sys/net/application_layer/nanocoap/sock.c:160]:
```suggestion
                if (cb) {
                    res = cb(arg, pkt);
                }
```

Maybe the callback could be optional?

File: sys/suit/handlers_command_seq.c [hunk 7]
```diff
@@ -359,9 +359,8 @@ static int _dtv_fetch(suit_manifest_t *manifest, int key,
     if (0) {}
 #ifdef MODULE_SUIT_TRANSPORT_COAP
     else if (strncmp(manifest->urlbuf, "coap://", 7) == 0) {
-        uint8_t buffer[NANOCOAP_BLOCKWISE_BUF(CONFIG_SUIT_COAP_BLOCKSIZE)];
```
Comment (maribu) [inline sys/suit/handlers_command_seq.c:362]: Care to sneak in a fix for the monstrosity in line 52?

**Review by maribu — COMMENTED** [review]
Code looks good to me. I think however that the callback should be called for each and any response, regardless of request type (see reasoning inline). However, one could make the callback function optional by allowing `NULL` is callback, so that the caller can decide if it makes sense to check the response (e.g. for the status code or options).

File: sys/net/application_layer/nanocoap/sock.c [hunk 8]
```diff
@@ -41,12 +54,38 @@ int nanocoap_sock_connect(nanocoap_sock_t *sock, sock_udp_ep_t *local, sock_udp_
...
+static bool _expect_response(const coap_pkt_t *pkt)
+{
+    switch (pkt->hdr->code) {
+    case COAP_METHOD_PUT:
+    case COAP_METHOD_DELETE:
+        return false;
+    }
+
+    return true;
+}
```
[the remark this answers is not in the document]
Comment (miri64) [reply sys/net/application_layer/nanocoap/sock.c:78]: > Wait I would get a response for a non-confirmable PUT?
> I thought the idea of PUT was that it is one-way 😨

You _always_ get a response for a CoAP request (if it comes through). Usually, PUT is used with RESTful APIs to create an object and POST to update it.

[the remark this answers is not in the document]
Comment (miri64) [reply sys/net/application_layer/nanocoap/sock.c:78]: Responses, however, are not as tightly bound to the classic request/response-paradigm. https://datatracker.ietf.org/doc/draft-bormann-core-responses/ gives an interesting read on that.

[the remark this answers is not in the document]
Comment (miri64) [reply sys/net/application_layer/nanocoap/sock.c:78]: > The callback will still be expected if the message was CONfirmable.
> 
> I assumed that when a) the message does not require a response and b) no response was requested, there will be no response.

Also: don't confuse layers: on the request/response layer there _always_ will be a response to a request. On the message layer, the expected message type back is an ACK (ACKs != responses, e.g. you can also have confirmable responses e.g. for an Observe notification or if the server responded with an empty ACK first, if it needs some longer processing times of the request for instance).

[the remark this answers is not in the document]
Comment (benpicco) [reply sys/net/application_layer/nanocoap/sock.c:78]: So if I have a continuous stream of measurement data where I don’t care if a single data point is lost (e.g. an audio stream) but I *do* care about latency and not disrupting a bus network by avoiding sending replies to each data point, I can’t use non-confirmable PUT?

[the remark this answers is not in the document]
Comment (miri64) [reply sys/net/application_layer/nanocoap/sock.c:78]: If your singular goal of using non-confirmable PUT is to not cause replies: [No](https://datatracker.ietf.org/doc/html/rfc7252#section-5.8.3). The [server may send a reply back](https://datatracker.ietf.org/doc/html/rfc7252#section-5.2.3). Why would you even use CoAP for such an application and not just e.g. raw UDP?

[the remark this answers is not in the document]
Comment (chrysn) [reply sys/net/application_layer/nanocoap/sock.c:78]: You can avoid sending replies by using the No-Response option in combination with NONs. However, beware that CoAP places rather strict limitations on the data rate you can send without any response (1 byte per second by default), which make it rather unsuitable for the audio stream case without closed-loop flow control established (which boils down to some form of lockstepping or a TCP-style congestion control). Going open-loop is not something the wider Internet supports; you sure can do it on local networks with static topologies, but even there it needs good evaluation and will not be simply caught by setting a single parameter to be suitable for generic 6lo networks.

Also, coming back to the PUT responses, they can provide valuable information to the client even in the success case. For example, PUTting data can result in an ETag on the response, which can be used later to PUT If-Match (avoiding overwriting another client's state).

[the remark this answers is not in the document]
Comment (benpicco) [reply sys/net/application_layer/nanocoap/sock.c]: Ok, changed it to always expect a response. #17544 will make it easier to check for `COAP_OPT_NO_RESPONSE` so I'm not adding this here.

[the remark this answers is not in the document]
Comment (maribu) [reply sys/net/application_layer/nanocoap/sock.c]: > Usually, PUT is used with RESTful APIs to create an object and POST to update it.

The other way around ;) `PUT` is idempotent, `POST` is not. So `PUT` usually has a "create and initialize with or update to" semantics (which would result in different response codes (Created or Changed), but the resulting state is identical no matter what). A "create new instance of" semantic has to be `POST`, as each request would create a new instance and it is not idempotent.

So `PUT` is used for idempotent updates (and may also create if non-existing), and `POST` is for creating or non-idempotent updates. (E.g. raise volume by 10 steps is a non-idempotent update, and can only be done with `POST`.)

[the remark this answers is not in the document]
Comment (miri64) [reply sys/net/application_layer/nanocoap/sock.c]: > The other way around ;) `PUT` is idempotent, `POST` is not. So `PUT` usually has a "create and initialize with or update to" semantics (which would result in different response codes (Created or Changed), but the resulting state is identical no matter what). A "create new instance of" semantic has to be `POST`, as each request would create a new instance and it is not idempotent.

You are right. Sorry, somehow mixed it up :sweat_smile:.

[the remark this answers is not in the document]
Comment (maribu) [reply sys/net/application_layer/nanocoap/sock.c]: At least you didn't mix it up in front of a room full of students that you have to teach it to. I get at least once every term a key detail wrong and have to explain the students that I f**ked up last week the week after :laughing:

File: sys/include/net/nanocoap.h (hunk 1 above)
Comment (benpicco) [reply sys/include/net/nanocoap.h:160]: Previously the work buffer was also used for the request.
Now that we don't need a work buffer anymore, we allocate the request header on the stack.
Turns out it needs to be a bit larger for e.g. SUIT. (previously this define was not used for enforcing a length restriction, but to calculate the requires user buffer size by doing `CONFIG_NANOCOAP_BLOCK_HEADER_MAX + BLOCKSIZE`.

Comment (maribu) [reply sys/include/net/nanocoap.h:160]: So maybe add

```
/* Value obtained experimentally when using SUIT */
```

in front, so that other developers how to check for regressions when changing this?

File: sys/net/application_layer/nanocoap/sock.c [hunk 9]
```diff
@@ -41,59 +55,111 @@ int nanocoap_sock_connect(nanocoap_sock_t *sock, sock_udp_ep_t *local, sock_udp_
...
+    if ((ep.family == AF_INET6) &&
+        ipv6_addr_is_multicast((ipv6_addr_t *)ep.addr.ipv6)) {
+        return false;
+    }
+#endif
+
+    /* TODO: check for COAP_OPT_NO_RESPONSE */
+
+    return true;
+}
```
[the remark this answers is not in the document]
Comment (benpicco) [reply sys/net/application_layer/nanocoap/sock.c]: True, with `cb` being optional now this can be handled much easier.

File: sys/net/application_layer/nanocoap/sock.c [hunk 10]
```diff
@@ -41,59 +55,90 @@ int nanocoap_sock_connect(nanocoap_sock_t *sock, sock_udp_ep_t *local, sock_udp_
...
     size_t pdu_len = (pkt->payload - (uint8_t *)pkt->hdr) + pkt->payload_len;
     uint8_t *buf = (uint8_t*)pkt->hdr;
     uint32_t id = coap_get_id(pkt);
+    void *payload, *ctx = NULL;
+
+    unsigned state = STATE_SEND_REQUEST;
 
-    /* TODO: timeout random between between ACK_TIMEOUT and (ACK_TIMEOUT *
-     * ACK_RANDOM_FACTOR) */
+    /* TODO: timeout random between between ACK_TIMEOUT and (ACK_TIMEOUT * ACK_RANDOM_FACTOR) */
```
Comment (maribu) [inline sys/net/application_layer/nanocoap/sock.c:92]: This one really should be addressed, but as it is pre-existing, I guess it deserves a dedicated PR.

**Comment by benpicco** [comment]
I did some tests:

<details><summary>examples/suit_update</summary>

```
suit: received URL: "coap://[2001:db8::1]/fw/same54-xpro/suit_update-riot.suit_signed.latest.bin"
suit_coap: trigger received
suit_coap: downloading "coap://[2001:db8::1]/fw/same54-xpro/suit_update-riot.suit_signed.latest.bin"
suit_coap: got manifest with size 495
suit: verifying manifest signature
suit: validated manifest version
)riotboot_hdr_validate: riotboot_hdr magic number invalid
Manifest seq_no: 1649865360, highest available: 1649865266
suit: validated sequence number
)Formatted component name: 
Comparing manifest offset 4000 with other slot offset
Comparing manifest offset 82000 with other slot offset
validating vendor ID
Comparing 547d0d74-6d3a-5a92-9662-4881afd9407b to 547d0d74-6d3a-5a92-9662-4881afd9407b from manifest
validating vendor ID: OK
validating class id
Comparing 50244518-6a7c-5ce7-932b-88b318336c82 to 50244518-6a7c-5ce7-932b-88b318336c82 from manifest
validating class id: OK
Comparing manifest offset 4000 with other slot offset
Comparing manifest offset 82000 with other slot offset
SUIT policy check OK.
Formatted component name: 
riotboot_flashwrite: initializing update to target slot 1
Fetching firmware |█████████████████████████| 100%
Finalizing payload store
Verifying image digest
Starting digest verification against image
Install correct payload
Verifying image digest
Starting digest verification against image
Install correct payload
Image magic_number: 0x544f4952
Image Version: 0x6256f290
Image start address: 0x00082400
Header chksum: 0x96c91691

suit_coap: rebooting...
----> ethos: hello received
Failed to send flush request: Operation not permitted
gnrc_uhcpc: Using 4 as border interface and 0 as wireless interface.
gnrc_uhcpc: only one interface found, skipping setup.
main(): This is RIOT! (Version: 2022.07-devel-16-g2dc98-nanocoap-rework)
RIOT SUIT update example application
Running from slot 1
Image magic_number: 0x544f4952
Image Version: 0x6256f290
Image start address: 0x00082400
Header chksum: 0x96c91691

suit_coap: started.
Starting the shell
> 
```
</details>

<details><summary>tests/nanocoap_cli</summary>

```
2022-04-13 19:09:23,212 # url get coap://coap.me/test
2022-04-13 19:09:23,243 # offset 000: welcome to the ETSI plugtest! la
2022-04-13 19:09:23,273 # offset 032: st change: 2022-04-13 17:08:55 U
2022-04-13 19:09:23,300 # offset 064: TC
2022-04-13 19:11:22,553 # > server start
2022-04-13 19:11:22,555 # starting server on port 5683
```

and on Linux

```
% aiocoap-client coap://[2001:16b8:45d0:db00:fec2:3dff:fe23:22df]/.well-known/core
application/link-format content was re-formatted
</.well-known/core>,
</value>

% aiocoap-client coap://[2001:16b8:45d0:db00:fec2:3dff:fe23:22df]/value 
0
```
</details>

<details><summary>examples/gcoap</summary>

```
2022-04-13 19:14:43,191 # >coap get coap.me 5683 /large
2022-04-13 19:14:43,195 # gcoap_cli: sending msg ID 64297, 12 bytes
2022-04-13 19:14:43,304 # gcoap: warning, incomplete response; continuing with the truncated payload
2022-04-13 19:14:43,306 # --- blockwise start ---
2022-04-13 19:14:43,311 # gcoap: response Success, code 2.05, 109 bytes
2022-04-13 19:14:43,311 # 
2022-04-13 19:14:43,316 #      0                   1                   2                   3
2022-04-13 19:14:43,320 #     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8
2022-04-13 19:14:43,353 # gcoap: warning, incomplete response; continuing with the truncated payload
2022-04-13 19:14:43,357 # gcoap: response Success, code 2.05, 109 bytes
2022-04-13 19:14:43,358 # 8 9 0 1
2022-04-13 19:14:43,364 #    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
2022-04-13 19:14:43,367 #    |Ver| T |  TKL  |      Code  
2022-04-13 19:14:43,399 # gcoap: warning, incomplete response; continuing with the truncated payload
2022-04-13 19:14:43,403 # gcoap: response Success, code 2.05, 109 bytes
2022-04-13 19:14:43,405 # ge ID           |
2022-04-13 19:14:43,411 #    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
2022-04-13 19:14:43,413 #    |   Token (if any, 
2022-04-13 19:14:43,445 # gcoap: warning, incomplete response; continuing with the truncated payload
2022-04-13 19:14:43,449 # gcoap: response Success, code 2.05, 109 bytes
2022-04-13 19:14:43,455 # -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
2022-04-13 19:14:43,457 #    |   Options (if any) ...
2022-04-13 19:14:43,459 #    +-+-+-+-+-+-+
2022-04-13 19:14:43,491 # gcoap: warning, incomplete response; continuing with the truncated payload
2022-04-13 19:14:43,495 # gcoap: response Success, code 2.05, 109 bytes
2022-04-13 19:14:43,498 # +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
2022-04-13 19:14:43,502 #    |1 1 1 1 1 1 1 1|    Payload (if any) ...
2022-04-13 19:14:43,505 #    +-+-+-+-+-+-+-+-+-+-+-+-+-+
2022-04-13 19:14:43,537 # gcoap: warning, incomplete response; continuing with the truncated payload
2022-04-13 19:14:43,541 # gcoap: response Success, code 2.05, 109 bytes
2022-04-13 19:14:43,543 # +-+-+-+-+-+-+-+-+-+
2022-04-13 19:14:43,543 # 
2022-04-13 19:14:43,544 # [...]
2022-04-13 19:14:43,544 # 
2022-04-13 19:14:43,550 #    Token Length (TKL):  4-bit unsigned integer.  Indicates the length of
2022-04-13 19:14:43,551 #       th
2022-04-13 19:14:43,583 # gcoap: warning, incomplete response; continuing with the truncated payload
2022-04-13 19:14:43,587 # gcoap: response Success, code 2.05, 109 bytes
2022-04-13 19:14:43,591 # oken field (0-8 bytes).  Lengths 9-15 are
2022-04-13 19:14:43,597 #       reserved, MUST NOT be sent, and MUST be processed as a messag
2022-04-13 19:14:43,630 # gcoap: warning, incomplete response; continuing with the truncated payload
2022-04-13 19:14:43,634 # gcoap: response Success, code 2.05, 109 bytes
2022-04-13 19:14:43,634 # r.
2022-04-13 19:14:43,634 # 
2022-04-13 19:14:43,640 #    Code:  8-bit unsigned integer, split into a 3-bit class (most
2022-04-13 19:14:43,643 #       significant bits) and a 5-bit deta
2022-04-13 19:14:43,676 # gcoap: warning, incomplete response; continuing with the truncated payload
2022-04-13 19:14:43,680 # gcoap: response Success, code 2.05, 109 bytes
2022-04-13 19:14:43,681 # nt bits),
2022-04-13 19:14:43,687 #       documented as c.dd where c is a digit from 0 to 7 for the 3-bit
2022-04-13 19:14:43,689 #       subfield and dd are two
2022-04-13 19:14:43,722 # gcoap: warning, incomplete response; continuing with the truncated payload
2022-04-13 19:14:43,726 # gcoap: response Success, code 2.05, 109 bytes
2022-04-13 19:14:43,728 # 31 for the 5-bit
2022-04-13 19:14:43,733 #       subfield.  The class can indicate a request (0), a success
2022-04-13 19:14:43,736 #       response (2), a clien
2022-04-13 19:14:43,768 # gcoap: warning, incomplete response; continuing with the truncated payload
2022-04-13 19:14:43,772 # gcoap: response Success, code 2.05, 109 bytes
2022-04-13 19:14:43,774 # ), or a server error
2022-04-13 19:14:43,780 #       response (5).  (All other class values are reserved.)  As a
2022-04-13 19:14:43,782 #       special case, Co
2022-04-13 19:14:43,815 # gcoap: warning, incomplete response; continuing with the truncated payload
2022-04-13 19:14:43,819 # gcoap: response Success, code 2.05, 109 bytes
2022-04-13 19:14:43,821 # n Empty message.  In case of a
2022-04-13 19:14:43,828 #       request, the Code field indicates the Request Method; in case of a
2022-04-13 19:14:43,828 #      
2022-04-13 19:14:43,860 # gcoap: warning, incomplete response; continuing with the truncated payload
2022-04-13 19:14:43,865 # gcoap: response Success, code 2.05, 109 bytes
2022-04-13 19:14:43,869 # e Code.  Possible values are maintained in the
2022-04-13 19:14:43,874 #       CoAP Code Registries (Section 12.1).  The semantics of r
2022-04-13 19:14:43,904 # gcoap: response Success, code 2.05, 36 bytes
2022-04-13 19:14:43,907 # esponses are defined in Section 5.
2022-04-13 19:14:43,907 # 
2022-04-13 19:14:43,908 # 
2022-04-13 19:14:43,910 # --- blockwise complete ---
```
</details>

**Review by maribu — APPROVED** [review]
ACK. Code looks good and I trust your testing
